## DS4A | Team 37 | Data Cleaning

Load necessary libraries

In [1]:
import pandas as pd
import numpy as np
import sklearn
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

In [2]:
import warnings            
warnings.filterwarnings("ignore")

In [3]:
import pandas as pd
import os

In [4]:
# Set `main_dir` to the location of your `skill-soft` folder (for Mac).
main_dir = '/Volumes/GoogleDrive/My Drive/DS4A/Project'

# Make `data_dir` from the `main_dir` and
# remainder of the path to data directory.
data_dir = main_dir + "/Data_Cleaning"

# Set working directory.
os.chdir(data_dir)
# Check working directory.
print(os.getcwd())

df = pd.read_csv('Hospital_Inpatient_Discharges__SPARCS_De-Identified___2017.csv')

/Volumes/GoogleDrive/My Drive/DS4A/Project/Data_Cleaning


Open data file and study the general info

In [5]:
df.info()
df.describe()
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2343569 entries, 0 to 2343568
Data columns (total 34 columns):
 #   Column                               Dtype  
---  ------                               -----  
 0   Hospital Service Area                object 
 1   Hospital County                      object 
 2   Operating Certificate Number         float64
 3   Permanent Facility Id                float64
 4   Facility Name                        object 
 5   Age Group                            object 
 6   Zip Code - 3 digits                  object 
 7   Gender                               object 
 8   Race                                 object 
 9   Ethnicity                            object 
 10  Length of Stay                       object 
 11  Type of Admission                    object 
 12  Patient Disposition                  object 
 13  Discharge Year                       int64  
 14  CCS Diagnosis Code                   int64  
 15  CCS Diagnosis Description       

Hospital Service Area                     5155
Hospital County                           5155
Operating Certificate Number              5155
Permanent Facility Id                     5155
Facility Name                                0
Age Group                                    0
Zip Code - 3 digits                      39019
Gender                                       0
Race                                         0
Ethnicity                                    0
Length of Stay                               0
Type of Admission                            0
Patient Disposition                          0
Discharge Year                               0
CCS Diagnosis Code                           0
CCS Diagnosis Description                    0
CCS Procedure Code                           0
CCS Procedure Description                    0
APR DRG Code                                 0
APR DRG Description                          0
APR MDC Code                                 0
APR MDC Descr

# 1. Deal with missing values

### Investigate columns that have missing values

Investigate why Birth Weight has such a great number of missing values

In [6]:
#Investigate why Birth Weight has such a great number of missing values
with pd.option_context('display.max_rows', None, 'display.max_columns', 34): 
    print(df[df['Birth Weight'].isnull()].head(10))

  Hospital Service Area Hospital County  Operating Certificate Number  \
0         Hudson Valley     Westchester                     5903001.0   
1         Hudson Valley     Westchester                     5903001.0   
2         Hudson Valley     Westchester                     5903001.0   
3         Hudson Valley     Westchester                     5903001.0   
4         Hudson Valley     Westchester                     5903001.0   
5         Hudson Valley     Westchester                     5903001.0   
6         Hudson Valley     Westchester                     5903001.0   
7         Hudson Valley     Westchester                     5903001.0   
8         Hudson Valley     Westchester                     5903001.0   
9         Hudson Valley     Westchester                     5903001.0   

   Permanent Facility Id                     Facility Name Age Group  \
0                 1061.0  Montefiore Mount Vernon Hospital  30 to 49   
1                 1061.0  Montefiore Mount Vernon Ho

In [7]:
a = df['CCS Diagnosis Code'].unique()
print(sorted(a))

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 2

In [8]:
df.loc[(df['Birth Weight'].isnull()) & (df['CCS Diagnosis Code'] == 11)]['CCS Diagnosis Code'].count()

2322

Investigate missng values in zip code

In [10]:
with pd.option_context('display.max_rows', None, 'display.max_columns', 34): 
    print(df[df['Zip Code - 3 digits'].isnull()].head(10))

    Hospital Service Area Hospital County  Operating Certificate Number  \
0           Hudson Valley     Westchester                     5903001.0   
25          Hudson Valley     Westchester                     5904001.0   
28                    NaN             NaN                           NaN   
61          Hudson Valley        Sullivan                     5263000.0   
213         New York City       Manhattan                     7002002.0   
225         New York City       Manhattan                     7002002.0   
277         New York City       Manhattan                     7002024.0   
405         New York City       Manhattan                     7002032.0   
426         New York City       Manhattan                     7002032.0   
471         New York City       Manhattan                     7002032.0   

     Permanent Facility Id                     Facility Name    Age Group  \
0                   1061.0  Montefiore Mount Vernon Hospital     30 to 49   
25                  

In [11]:
df[df['Facility Name']== 'Montefiore Mount Vernon Hospital'].head(5)

Hospital Service Area Hospital County  Operating Certificate Number  \
0         Hudson Valley     Westchester                     5903001.0   
1         Hudson Valley     Westchester                     5903001.0   
2         Hudson Valley     Westchester                     5903001.0   
3         Hudson Valley     Westchester                     5903001.0   
4         Hudson Valley     Westchester                     5903001.0   

   Permanent Facility Id                     Facility Name Age Group  \
0                 1061.0  Montefiore Mount Vernon Hospital  30 to 49   
1                 1061.0  Montefiore Mount Vernon Hospital  50 to 69   
2                 1061.0  Montefiore Mount Vernon Hospital  30 to 49   
3                 1061.0  Montefiore Mount Vernon Hospital  50 to 69   
4                 1061.0  Montefiore Mount Vernon Hospital  18 to 29   

  Zip Code - 3 digits Gender   Race          Ethnicity  ...  \
0                 NaN      M  White  Not Span/Hispanic  ...   
1                 105      M  White   Spanish/Hispanic  ...   
2                 105      F  White            Unknown  ...   
3                 105      F  White  Not Span/Hispanic  ...   
4                 105      F  White   Spanish/Hispanic  ...   

  APR Risk of Mortality APR Medical Surgical Description  \
0                 Minor                          Medical   
1              Moderate                          Medical   
2              Moderate                          Medical   
3              Moderate                          Medical   
4                 Minor                          Medical   

          Payment Typology 1  Payment Typology 2  Payment Typology 3  \
0                   Medicare            Medicare            Self-Pay   
1                   Medicaid            Medicaid            Self-Pay   
2  Department of Corrections            Self-Pay                 NaN   
3                   Medicaid            Medicaid            Self-Pay   
4  Department of Corrections            Self-Pay                 NaN   

  Birth Weight  Abortion Edit Indicator Emergency Department Indicator  \
0          NaN                        N                              N   
1          NaN                        N                              N   
2          NaN                        N                              N   
3          NaN                        N                              N   
4          NaN                        N                              N   

   Total Charges Total Costs  
0      114168.00    40464.04  
1       58517.00    26046.50  
2       39647.04    17397.59  
3       36031.09    16118.84  
4       30534.00    11921.05  

[5 rows x 34 columns]

### Recover missing zip codes from the same hospital 

- Create a dictionary that has info on hospital id and zip code (extract rows from the current data, remove duplicate rows)
- Merge with the current dataset 

In [12]:
zip_code = df[['Permanent Facility Id', 'Operating Certificate Number','Facility Name', 'Zip Code - 3 digits']]

In [13]:
print('Number of unique "Permanent Facility Id":',len(df['Permanent Facility Id'].unique()))
print('Number of unique "Operating Certificate Number":',len(df['Operating Certificate Number'].unique()))
print('Number of unique "Facility Name":',len(df['Facility Name'].unique()))

Number of unique "Permanent Facility Id": 212
Number of unique "Operating Certificate Number": 175
Number of unique "Facility Name": 211


In [14]:
#Remove rows with missing value
zip_code_clean = zip_code.dropna(subset=['Zip Code - 3 digits'])
zip_code_clean.isna().sum()

Permanent Facility Id           0
Operating Certificate Number    0
Facility Name                   0
Zip Code - 3 digits             0
dtype: int64

In [15]:
zip_code_clean_1 = zip_code_clean.drop_duplicates(subset=['Permanent Facility Id'])

zip_code_clean_1[zip_code_clean_1.duplicated()].head() #Done

zip_code_clean_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 210 entries, 1 to 1125696
Data columns (total 4 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Permanent Facility Id         210 non-null    float64
 1   Operating Certificate Number  210 non-null    float64
 2   Facility Name                 210 non-null    object 
 3   Zip Code - 3 digits           210 non-null    object 
dtypes: float64(2), object(2)
memory usage: 8.2+ KB


In [16]:
df_full_zipcode = pd.merge(df, zip_code_clean_1, how='left', on='Permanent Facility Id')

df_full_zipcode.isna().sum()

Hospital Service Area                     5155
Hospital County                           5155
Operating Certificate Number_x            5155
Permanent Facility Id                     5155
Facility Name_x                              0
Age Group                                    0
Zip Code - 3 digits_x                    39019
Gender                                       0
Race                                         0
Ethnicity                                    0
Length of Stay                               0
Type of Admission                            0
Patient Disposition                          0
Discharge Year                               0
CCS Diagnosis Code                           0
CCS Diagnosis Description                    0
CCS Procedure Code                           0
CCS Procedure Description                    0
APR DRG Code                                 0
APR DRG Description                          0
APR MDC Code                                 0
APR MDC Descr

There are still missing zip codes, let investigate why

In [17]:
df_full_zipcode[df_full_zipcode['Zip Code - 3 digits_y'].isnull()]['Facility Name_x'].unique()

array(['Redacted for Confidentiality',
       'Albany Medical Center - South Clinical Campus'], dtype=object)

Because we dont have zip code for 
- Albany Medical Center - South Clinical Campus: this can be googled 
- Redacted for Confidentiality: we can't do anything about this

In [18]:
#Have a look at the Albany Medical Center - South Clinical Campus
df[df['Facility Name'] == 'Albany Medical Center - South Clinical Campus']

Hospital Service Area Hospital County  Operating Certificate Number  \
485306        Capital/Adirond          Albany                      101000.0   
485307        Capital/Adirond          Albany                      101000.0   
800703        Capital/Adirond          Albany                      101000.0   
1134235       Capital/Adirond          Albany                      101000.0   
1134236       Capital/Adirond          Albany                      101000.0   
1637241       Capital/Adirond          Albany                      101000.0   

         Permanent Facility Id                                  Facility Name  \
485306                     2.0  Albany Medical Center - South Clinical Campus   
485307                     2.0  Albany Medical Center - South Clinical Campus   
800703                     2.0  Albany Medical Center - South Clinical Campus   
1134235                    2.0  Albany Medical Center - South Clinical Campus   
1134236                    2.0  Albany Medical Center - South Clinical Campus   
1637241                    2.0  Albany Medical Center - South Clinical Campus   

        Age Group Zip Code - 3 digits Gender   Race          Ethnicity  ...  \
485306   50 to 69                 NaN      F  White  Not Span/Hispanic  ...   
485307   50 to 69                 NaN      F  White  Not Span/Hispanic  ...   
800703   50 to 69                 NaN      F  White  Not Span/Hispanic  ...   
1134235  50 to 69                 NaN      F  White  Not Span/Hispanic  ...   
1134236  18 to 29                 NaN      F  White  Not Span/Hispanic  ...   
1637241  18 to 29                 NaN      F  White  Not Span/Hispanic  ...   

        APR Risk of Mortality APR Medical Surgical Description  \
485306                  Minor                         Surgical   
485307               Moderate                         Surgical   
800703                  Minor                         Surgical   
1134235              Moderate                         Surgical   
1134236                 Minor                          Medical   
1637241                 Minor                         Surgical   

             Payment Typology 1  Payment Typology 2  Payment Typology 3  \
485306   Blue Cross/Blue Shield                 NaN                 NaN   
485307                 Medicare                 NaN                 NaN   
800703   Blue Cross/Blue Shield                 NaN                 NaN   
1134235  Blue Cross/Blue Shield                 NaN                 NaN   
1134236                Medicare                 NaN                 NaN   
1637241  Blue Cross/Blue Shield            Medicare                 NaN   

        Birth Weight  Abortion Edit Indicator Emergency Department Indicator  \
485306           NaN                        N                              N   
485307           NaN                        N                              N   
800703           NaN                        N                              N   
1134235          NaN                        N                              N   
1134236          NaN                        N                              N   
1637241          NaN                        N                              N   

         Total Charges Total Costs  
485306        15200.29     5929.01  
485307        22527.77     6347.42  
800703        36810.92    13565.10  
1134235       12613.10     4986.62  
1134236        8976.79     3738.74  
1637241       11143.94     5074.75  

[6 rows x 34 columns]

In [19]:
zip_code_clean_1.head()

Permanent Facility Id  Operating Certificate Number  \
1                   1061.0                     5903001.0   
11                  1072.0                     5904001.0   
30                  1439.0                     7002002.0   
32                   971.0                     5263000.0   
136                 1324.0                     7002002.0   

                        Facility Name Zip Code - 3 digits  
1    Montefiore Mount Vernon Hospital                 105  
11   Montefiore New Rochelle Hospital                 108  
30            Mount Sinai Beth Israel                 100  
32   Catskill Regional Medical Center                 127  
136              Mount Sinai Brooklyn                 112

Add zip code of Albany Medical Center - South Clinical Campus

In [20]:
albany = pd.DataFrame([[2.0, 101000.0, 'Albany Medical Center - South Clinical Campus', 122]], 
                      columns=['Permanent Facility Id', 'Operating Certificate Number', 'Facility Name', 'Zip Code - 3 digits'])

zip_code_clean_2 = zip_code_clean_1.append(albany)

zip_code_clean_2.tail()

Permanent Facility Id  Operating Certificate Number  \
702780                   114.0                      427000.0   
707558                   804.0                     4402000.0   
1010230                 1005.0                     5601000.0   
1125696                  414.0                     2701006.0   
0                          2.0                      101000.0   

                                         Facility Name Zip Code - 3 digits  
702780          TLC Health Network Lake Shore Hospital                 141  
707558                       Massena Memorial Hospital                 136  
1010230                           Glens Falls Hospital                 128  
1125696                      Monroe Community Hospital                 146  
0        Albany Medical Center - South Clinical Campus                 122

Merge with the current data to get all of the zip code we can recover, except for the redacted ones

In [21]:
df_full_zipcode_1 = pd.merge(df, zip_code_clean_2, how='left', on='Permanent Facility Id')\
    .drop(['Operating Certificate Number_y','Facility Name_y', 'Zip Code - 3 digits_x'], axis=1)\
    .rename(columns={"Zip Code - 3 digits_y": "Zip Code - 3 digits", 
                     "Operating Certificate Number_x":"Operating Certificate Number",
                    "Facility Name_x": "Facility Name"})
        

In [22]:
df_full_zipcode_1.isna().sum()


Hospital Service Area                     5155
Hospital County                           5155
Operating Certificate Number              5155
Permanent Facility Id                     5155
Facility Name                                0
Age Group                                    0
Gender                                       0
Race                                         0
Ethnicity                                    0
Length of Stay                               0
Type of Admission                            0
Patient Disposition                          0
Discharge Year                               0
CCS Diagnosis Code                           0
CCS Diagnosis Description                    0
CCS Procedure Code                           0
CCS Procedure Description                    0
APR DRG Code                                 0
APR DRG Description                          0
APR MDC Code                                 0
APR MDC Description                          0
APR Severity 

In [23]:
df_full_zipcode_1[df_full_zipcode_1['Hospital Service Area'].isnull()].head()

Hospital Service Area Hospital County  Operating Certificate Number  \
28                     NaN             NaN                           NaN   
595                    NaN             NaN                           NaN   
839                    NaN             NaN                           NaN   
2227                   NaN             NaN                           NaN   
2723                   NaN             NaN                           NaN   

      Permanent Facility Id                 Facility Name Age Group Gender  \
28                      NaN  Redacted for Confidentiality   0 to 17      F   
595                     NaN  Redacted for Confidentiality  18 to 29      F   
839                     NaN  Redacted for Confidentiality  18 to 29      F   
2227                    NaN  Redacted for Confidentiality   0 to 17      F   
2723                    NaN  Redacted for Confidentiality  30 to 49      F   

                        Race          Ethnicity Length of Stay  ...  \
28                     White  Not Span/Hispanic              2  ...   
595                    White  Not Span/Hispanic              1  ...   
839   Black/African American  Not Span/Hispanic              1  ...   
2227              Other Race   Spanish/Hispanic              1  ...   
2723  Black/African American  Not Span/Hispanic              1  ...   

     APR Medical Surgical Description      Payment Typology 1  \
28                           Surgical  Blue Cross/Blue Shield   
595                           Medical  Blue Cross/Blue Shield   
839                          Surgical                Medicaid   
2227                          Medical                Medicaid   
2723                         Surgical                Medicaid   

      Payment Typology 2  Payment Typology 3 Birth Weight  \
28              Self-Pay                 NaN          NaN   
595             Medicaid                 NaN          NaN   
839             Medicaid            Self-Pay          NaN   
2227                 NaN                 NaN          NaN   
2723                 NaN                 NaN          NaN   

      Abortion Edit Indicator Emergency Department Indicator  Total Charges  \
28                          N                              N       39565.96   
595                         Y                              N        7089.08   
839                         Y                              N       17850.46   
2227                        Y                              Y       21040.48   
2723                        Y                              Y       30815.99   

     Total Costs  Zip Code - 3 digits  
28      15953.31                  NaN  
595      3101.13                  NaN  
839      7671.21                  NaN  
2227     4112.07                  NaN  
2723     6028.16                  NaN  

[5 rows x 34 columns]

### Remove rows that have NAs in the two columns that has a small number of NAs (< 0.01%)

In [24]:
df_clean = df_full_zipcode_1.dropna(subset=['APR Severity of Illness Description', 'APR Risk of Mortality'])
df_clean.isna().sum()
df_clean.dtypes

Hospital Service Area                   object
Hospital County                         object
Operating Certificate Number           float64
Permanent Facility Id                  float64
Facility Name                           object
Age Group                               object
Gender                                  object
Race                                    object
Ethnicity                               object
Length of Stay                          object
Type of Admission                       object
Patient Disposition                     object
Discharge Year                           int64
CCS Diagnosis Code                       int64
CCS Diagnosis Description               object
CCS Procedure Code                       int64
CCS Procedure Description               object
APR DRG Code                             int64
APR DRG Description                     object
APR MDC Code                             int64
APR MDC Description                     object
APR Severity 

# 2. Deal with incorrect data type

'Length of Stay' should be numeric, but it's currently an object, let's have a look at the variable

In [25]:
df_clean['Length of Stay'].unique()

array(['21', '8', '6', '4', '3', '1', '27', '14', '5', '2', '26', '24',
       '25', '20', '12', '11', '7', '9', '13', '10', '62', '60', '42',
       '36', '29', '28', '15', '16', '17', '44', '22', '64', '37', '19',
       '78', '65', '39', '35', '118', '68', '49', '30', '34', '18', '50',
       '56', '61', '41', '45', '51', '47', '31', '23', '80', '52', '38',
       '32', '113', '120 +', '71', '66', '48', '96', '54', '43', '91',
       '67', '46', '90', '33', '57', '74', '88', '81', '83', '75', '55',
       '107', '77', '72', '70', '59', '103', '110', '76', '40', '101',
       '73', '69', '53', '58', '98', '63', '106', '84', '85', '79', '108',
       '97', '92', '102', '99', '109', '94', '117', '105', '114', '104',
       '82', '112', '100', '116', '86', '119', '95', '87', '89', '93',
       '111', '115'], dtype=object)

Because there is "120+" value --> object. Let treat "120+" as 120 and convert it to numeric

In [26]:
df_clean['Length of Stay'] = df_clean['Length of Stay'].apply(lambda x: str(x).split(' ')[0])
df_clean['Length of Stay'] = pd.to_numeric(df_clean['Length of Stay'])
df_clean['Length of Stay'].unique()

array([ 21,   8,   6,   4,   3,   1,  27,  14,   5,   2,  26,  24,  25,
        20,  12,  11,   7,   9,  13,  10,  62,  60,  42,  36,  29,  28,
        15,  16,  17,  44,  22,  64,  37,  19,  78,  65,  39,  35, 118,
        68,  49,  30,  34,  18,  50,  56,  61,  41,  45,  51,  47,  31,
        23,  80,  52,  38,  32, 113, 120,  71,  66,  48,  96,  54,  43,
        91,  67,  46,  90,  33,  57,  74,  88,  81,  83,  75,  55, 107,
        77,  72,  70,  59, 103, 110,  76,  40, 101,  73,  69,  53,  58,
        98,  63, 106,  84,  85,  79, 108,  97,  92, 102,  99, 109,  94,
       117, 105, 114, 104,  82, 112, 100, 116,  86, 119,  95,  87,  89,
        93, 111, 115])

Birth Weight should be numeric, but it's an object. Let's look at the variable

In [29]:
df_clean['Birth Weight'].unique()

array([nan, 3300.0, 3100.0, 3000.0, 3400.0, 2500.0, 3800.0, 2300.0,
       2900.0, 3600.0, 1300.0, 2400.0, 2100.0, 2200.0, 1800.0, 3700.0,
       3200.0, 4000.0, 3500.0, 800.0, 2600.0, 2800.0, 1500.0, 700.0,
       5800.0, 1600.0, 2000.0, 3900.0, 4400.0, 4100.0, 4300.0, 4800.0,
       2700.0, 1900.0, 4200.0, 4500.0, 900.0, 4700.0, 500.0, 1700.0,
       1100.0, 4600.0, 400.0, 600.0, 1400.0, 5100.0, 300.0, 1000.0,
       1200.0, 5200.0, 4900.0, 5000.0, 5500.0, 5400.0, 200.0, 8800.0,
       100.0, 5300.0, 6000.0, 7000.0, 7800.0, 9000.0, 6800.0, 8000.0,
       6600.0, 5600.0, 7400.0, 8900.0, 8200.0, 6400.0, 6200.0, 8100.0,
       '04000', '03200', '03100', '03700', '03300', '03900', '02400',
       '03000', '02900', '03400', '02700', '03500', '02600', '03800',
       '03600', '00500', '02500', '04100', '04300', '02800', '00700',
       '02300', '01900', '02000', '02200', '04800', '04200', '01500',
       '00300', 'UNKN', '01600', '04400', '01400', '02100', '00800',
       '01300', '04700',

Because it contains "UNKN" and also weird presentation of number, like 04000. We can:
- Convert UNKN (Unknown) to NULL
- Convert all other to numeric


In [30]:
df_clean_1 = df_clean
df_clean_1['Birth Weight_clean'] = df_clean_1['Birth Weight'].apply(pd.to_numeric, errors='coerce')
df_clean_1['Birth Weight_clean'].unique()

array([   nan,  3300.,  3100.,  3000.,  3400.,  2500.,  3800.,  2300.,
        2900.,  3600.,  1300.,  2400.,  2100.,  2200.,  1800.,  3700.,
        3200.,  4000.,  3500.,   800.,  2600.,  2800.,  1500.,   700.,
        5800.,  1600.,  2000.,  3900.,  4400.,  4100.,  4300.,  4800.,
        2700.,  1900.,  4200.,  4500.,   900.,  4700.,   500.,  1700.,
        1100.,  4600.,   400.,   600.,  1400.,  5100.,   300.,  1000.,
        1200.,  5200.,  4900.,  5000.,  5500.,  5400.,   200.,  8800.,
         100.,  5300.,  6000.,  7000.,  7800.,  9000.,  6800.,  8000.,
        6600.,  5600.,  7400.,  8900.,  8200.,  6400.,  6200.,  8100.,
        6100.,  5700.,  6300.,  6900.,  8300., 11700.])

In [31]:
df_clean_1[df_clean_1['Birth Weight'] == 'UNKN']
df_clean_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2343329 entries, 0 to 2343568
Data columns (total 35 columns):
 #   Column                               Dtype  
---  ------                               -----  
 0   Hospital Service Area                object 
 1   Hospital County                      object 
 2   Operating Certificate Number         float64
 3   Permanent Facility Id                float64
 4   Facility Name                        object 
 5   Age Group                            object 
 6   Gender                               object 
 7   Race                                 object 
 8   Ethnicity                            object 
 9   Length of Stay                       int64  
 10  Type of Admission                    object 
 11  Patient Disposition                  object 
 12  Discharge Year                       int64  
 13  CCS Diagnosis Code                   int64  
 14  CCS Diagnosis Description            object 
 15  CCS Procedure Code              

In [32]:
#Rename Birth Weight_clean column to "Birth Weight" and drop the original column
df_clean_2 = df_clean_1.drop(['Birth Weight'], axis=1).rename(columns={"Birth Weight_clean": "Birth Weight"})
df_clean_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2343329 entries, 0 to 2343568
Data columns (total 34 columns):
 #   Column                               Dtype  
---  ------                               -----  
 0   Hospital Service Area                object 
 1   Hospital County                      object 
 2   Operating Certificate Number         float64
 3   Permanent Facility Id                float64
 4   Facility Name                        object 
 5   Age Group                            object 
 6   Gender                               object 
 7   Race                                 object 
 8   Ethnicity                            object 
 9   Length of Stay                       int64  
 10  Type of Admission                    object 
 11  Patient Disposition                  object 
 12  Discharge Year                       int64  
 13  CCS Diagnosis Code                   int64  
 14  CCS Diagnosis Description            object 
 15  CCS Procedure Code              

Investigate potential categorical variables

In [27]:
df_clean['APR Risk of Mortality'].value_counts()

Minor       1338452
Moderate     508068
Major        378128
Extreme      118681
Name: APR Risk of Mortality, dtype: int64

In [28]:
df_clean['Payment Typology 1'].value_counts()

Medicare                     906899
Medicaid                     703996
Private Health Insurance     344913
Blue Cross/Blue Shield       262808
Self-Pay                      44657
Managed Care, Unspecified     32315
Miscellaneous/Other           28515
Federal/State/Local/VA        14298
Department of Corrections      2622
Unknown                        2306
Name: Payment Typology 1, dtype: int64

Create new columns to transform categorical variables

In [33]:
age_dict = {'0 to 17': 1, '18 to 29': 2, '30 to 49': 3, '50 to 69': 4, '70 or Older': 5}

mortality_dict = {'Minor': 1, 'Moderate': 2, 'Major': 3, 'Extreme': 4}

payment_typo_dict = {'Medicare': 1, 'Medicaid': 2, 'Private Health Insurance': 3, 
                     'Blue Cross/Blue Shield': 4, 'Self-Pay': 5, 'Managed Care, Unspecified': 6, 
                     'Miscellaneous/Other': 7, 'Federal/State/Local/VA': 8, 
                     'Department of Corrections': 9, 'Unknown': 9999}

apr_medical_surgical_code = {'Medical': 1, 'Surgical': 2}

df_clean_2['age_group_code'] = df_clean_2['Age Group'].apply(lambda x: age_dict[x])
df_clean_2['apr_risk_mortality_code'] = df_clean_2['APR Risk of Mortality'].apply(lambda x: mortality_dict[x])
df_clean_2['payment_typology_1_code'] = df_clean_2['Payment Typology 1'].apply(lambda x: payment_typo_dict[x])
df_clean_2['apr_medical_surgical_code'] = df_clean_2['APR Medical Surgical Description'].\
                    apply(lambda x: apr_medical_surgical_code[x])


df_clean_2.head()

Hospital Service Area Hospital County  Operating Certificate Number  \
0         Hudson Valley     Westchester                     5903001.0   
1         Hudson Valley     Westchester                     5903001.0   
2         Hudson Valley     Westchester                     5903001.0   
3         Hudson Valley     Westchester                     5903001.0   
4         Hudson Valley     Westchester                     5903001.0   

   Permanent Facility Id                     Facility Name Age Group Gender  \
0                 1061.0  Montefiore Mount Vernon Hospital  30 to 49      M   
1                 1061.0  Montefiore Mount Vernon Hospital  50 to 69      M   
2                 1061.0  Montefiore Mount Vernon Hospital  30 to 49      F   
3                 1061.0  Montefiore Mount Vernon Hospital  50 to 69      F   
4                 1061.0  Montefiore Mount Vernon Hospital  18 to 29      F   

    Race          Ethnicity  Length of Stay  ... Abortion Edit Indicator  \
0  White  Not Span/Hispanic              21  ...                       N   
1  White   Spanish/Hispanic               8  ...                       N   
2  White            Unknown               6  ...                       N   
3  White  Not Span/Hispanic               4  ...                       N   
4  White   Spanish/Hispanic               4  ...                       N   

  Emergency Department Indicator  Total Charges  Total Costs  \
0                              N      114168.00     40464.04   
1                              N       58517.00     26046.50   
2                              N       39647.04     17397.59   
3                              N       36031.09     16118.84   
4                              N       30534.00     11921.05   

  Zip Code - 3 digits  Birth Weight age_group_code  apr_risk_mortality_code  \
0                 105           NaN              3                        1   
1                 105           NaN              4                        2   
2                 105           NaN              3                        2   
3                 105           NaN              4                        2   
4                 105           NaN              2                        1   

  payment_typology_1_code  apr_medical_surgical_code  
0                       1                          1  
1                       2                          1  
2                       9                          1  
3                       2                          1  
4                       9                          1  

[5 rows x 38 columns]

In [34]:
df_clean_2.isna().sum()

Hospital Service Area                     5150
Hospital County                           5150
Operating Certificate Number              5150
Permanent Facility Id                     5150
Facility Name                                0
Age Group                                    0
Gender                                       0
Race                                         0
Ethnicity                                    0
Length of Stay                               0
Type of Admission                            0
Patient Disposition                          0
Discharge Year                               0
CCS Diagnosis Code                           0
CCS Diagnosis Description                    0
CCS Procedure Code                           0
CCS Procedure Description                    0
APR DRG Code                                 0
APR DRG Description                          0
APR MDC Code                                 0
APR MDC Description                          0
APR Severity 

Transform from numeric to string 

In [35]:
df_clean_2[["Operating Certificate Number", "Permanent Facility Id"]] = df_clean_2[["Operating Certificate Number", 
                                                                                    "Permanent Facility Id"]].astype(str) 
df_clean_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2343329 entries, 0 to 2343568
Data columns (total 38 columns):
 #   Column                               Dtype  
---  ------                               -----  
 0   Hospital Service Area                object 
 1   Hospital County                      object 
 2   Operating Certificate Number         object 
 3   Permanent Facility Id                object 
 4   Facility Name                        object 
 5   Age Group                            object 
 6   Gender                               object 
 7   Race                                 object 
 8   Ethnicity                            object 
 9   Length of Stay                       int64  
 10  Type of Admission                    object 
 11  Patient Disposition                  object 
 12  Discharge Year                       int64  
 13  CCS Diagnosis Code                   int64  
 14  CCS Diagnosis Description            object 
 15  CCS Procedure Code              

### Prepare the data for modeling purpose
- Remove rows with NAs 
- Replace OOS character in the zip code (out of state) with numeric variable 9999
- Remove all unnecessary columns
- Transform all should-have-been-categorical variables to category

In [36]:
#Remove rows with NAs 
df_modeling = df_clean_2.dropna(subset=['Hospital Service Area'])

In [37]:
#Replace OOS character in the zip code (out of state) with numeric variable 9999
df_modeling['Zip Code - 3 digits_clean'] = df_modeling['Zip Code - 3 digits'].apply(lambda x: str(x).replace('OOS','9999'))  
df_modeling['Zip Code - 3 digits_clean'] = pd.to_numeric(df_modeling['Zip Code - 3 digits_clean'])

In [38]:
#Transform all should-have-been-categorical variables to category
df_modeling['CCS Diagnosis Code'] = df_modeling['CCS Diagnosis Code'].astype('category')
df_modeling['CCS Procedure Code'] = df_modeling['CCS Procedure Code'].astype('category')
df_modeling['APR DRG Code'] = df_modeling['APR DRG Code'].astype('category')
df_modeling['APR MDC Code'] = df_modeling['APR MDC Code'].astype('category')
df_modeling['APR Severity of Illness Code'] = df_modeling['APR Severity of Illness Code'].astype('category')
df_modeling['age_group_code'] = df_modeling['age_group_code'].astype('category')
df_modeling['apr_risk_mortality_code'] = df_modeling['apr_risk_mortality_code'].astype('category')
df_modeling['payment_typology_1_code'] = df_modeling['payment_typology_1_code'].astype('category')
df_modeling['apr_medical_surgical_code'] = df_modeling['apr_medical_surgical_code'].astype('category')

In [39]:
df_modeling.info()
df_modeling.isna().sum()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2338179 entries, 0 to 2343568
Data columns (total 39 columns):
 #   Column                               Dtype   
---  ------                               -----   
 0   Hospital Service Area                object  
 1   Hospital County                      object  
 2   Operating Certificate Number         object  
 3   Permanent Facility Id                object  
 4   Facility Name                        object  
 5   Age Group                            object  
 6   Gender                               object  
 7   Race                                 object  
 8   Ethnicity                            object  
 9   Length of Stay                       int64   
 10  Type of Admission                    object  
 11  Patient Disposition                  object  
 12  Discharge Year                       int64   
 13  CCS Diagnosis Code                   category
 14  CCS Diagnosis Description            object  
 15  CCS Procedure C

Hospital Service Area                        0
Hospital County                              0
Operating Certificate Number                 0
Permanent Facility Id                        0
Facility Name                                0
Age Group                                    0
Gender                                       0
Race                                         0
Ethnicity                                    0
Length of Stay                               0
Type of Admission                            0
Patient Disposition                          0
Discharge Year                               0
CCS Diagnosis Code                           0
CCS Diagnosis Description                    0
CCS Procedure Code                           0
CCS Procedure Description                    0
APR DRG Code                                 0
APR DRG Description                          0
APR MDC Code                                 0
APR MDC Description                          0
APR Severity 

In [40]:
df_modeling.to_csv('df_modeling.csv')
df_modeling.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2338179 entries, 0 to 2343568
Data columns (total 39 columns):
 #   Column                               Dtype   
---  ------                               -----   
 0   Hospital Service Area                object  
 1   Hospital County                      object  
 2   Operating Certificate Number         object  
 3   Permanent Facility Id                object  
 4   Facility Name                        object  
 5   Age Group                            object  
 6   Gender                               object  
 7   Race                                 object  
 8   Ethnicity                            object  
 9   Length of Stay                       int64   
 10  Type of Admission                    object  
 11  Patient Disposition                  object  
 12  Discharge Year                       int64   
 13  CCS Diagnosis Code                   category
 14  CCS Diagnosis Description            object  
 15  CCS Procedure C

In [43]:
#Check duplicate rows
#df_modeling[df_modeling.duplicated(keep = False)].head(10)

#Remove columns that have the same meaning
#Investigate these columns

print(len(df_modeling['APR Severity of Illness Code'].unique()))
print(len(df_modeling['APR Severity of Illness Description'].unique()))

df_modeling['APR Medical Surgical Description'].unique()

#Remove duplicate columns
df_modeling_clean = df_modeling.drop(['CCS Diagnosis Description', 'CCS Procedure Description',
                                     'APR DRG Description', 'APR MDC Description',
                                     'APR Severity of Illness Description',
                                     'APR Risk of Mortality', 'Payment Typology 1', 
                                      'APR Medical Surgical Description'], axis = 1)
df_modeling_clean.info()


4
4
<class 'pandas.core.frame.DataFrame'>
Int64Index: 2338179 entries, 0 to 2343568
Data columns (total 31 columns):
 #   Column                          Dtype   
---  ------                          -----   
 0   Hospital Service Area           object  
 1   Hospital County                 object  
 2   Operating Certificate Number    object  
 3   Permanent Facility Id           object  
 4   Facility Name                   object  
 5   Age Group                       object  
 6   Gender                          object  
 7   Race                            object  
 8   Ethnicity                       object  
 9   Length of Stay                  int64   
 10  Type of Admission               object  
 11  Patient Disposition             object  
 12  Discharge Year                  int64   
 13  CCS Diagnosis Code              category
 14  CCS Procedure Code              category
 15  APR DRG Code                    category
 16  APR MDC Code                    category
 17  APR 

In [ ]:
#Export modeling data
df_modeling_clean.to_csv('df_modeling_1.csv')

In [ ]:
#Export general data
df_clean_2.info()
df_clean_2.to_csv('df_general.csv')

%history -g